internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -166,6 +139,54 @@ func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
 	return <-errc
 }
 
+// Read from the channels established with NotifyPublish and NotifyReturn.
+// Must be called with t.mu held.
+func (t *topic) receiveFromPublishChannels(ctx context.Context, nMessages int) error {
+	// Consume all the acknowledgments for the messages we are publishing.
+	// Since SendBatch (the only caller of this method) holds the lock, we expect
+	// exactly as many acks as messages.
+	var err error
+	nAcks := 0
+	for nAcks < nMessages {
+		select {
+		case <-ctx.Done():
+			// Channel will be in a weird state (not all publish acks consumed, perhaps)
+			// so re-create it next time.
+			t.ch = nil
+			return ctx.Err()
+
+		case ret, ok := <-t.retc:
+			if !ok {
+				// Channel closed. Handled in the pubc case below. But set
+				// the channel to nil to prevent it from being selected again.
+				t.retc = nil
+			} else if err == nil {
+				// The message was returned from the server because it is unroutable.
+				// This will be the error we return, but continue so we drain all
+				// items from pubc. We don't need to re-establish the channel on this
+				// error.
+				err = fmt.Errorf("rabbitpubsub: message returned from %s: %s (code %d)",
+					ret.Exchange, ret.ReplyText, ret.ReplyCode)
```

Comment:
Don't you need to increment `nAcks` here?

Reply:
No. The contract is that I will get exactly one item on pubc for each message I send, whether there's a problem or not, but I'll only get something on retc for returned messages.

Reply:
Interesting. Can you add a comment explaining this? It's not obvious; I assumed you'd either get a `pubc` message OR a `retc` message, and then outer loop of `nAcks < nMessages` looks like a bug. Maybe at the top of the loop?

Reply:
Rewrote the comment just below the func line.